primitives/staking/src/offence.rs

```
@@ -209,3 +209,74 @@ pub struct OffenceDetails<Reporter, Offender> {
 	/// particular reporters.
 	pub reporters: Vec<Reporter>,
 }
+
+/// A system capable of construct, report and submit offences.
+///
+/// Implementors of this trait provide a wrapper for lower level operations.
+///
+/// It is assumed that this subsystem takes care of checking key ownership proof
+/// before report submission.
+pub trait OffenceReportSystem<Reporter, OffenceProof> {
+	/// Offender key ownership proof.
+	/// This can be used by the offence report system to check for evidence validity.
+	type KeyOwnerProof;
+
+	/// Longevity, in blocks, for the report validity. When using the staking
+	/// pallet this should be equal to the bonding duration (in blocks, not eras).
+	type ReportLongevity: Get<u64>;
+
+	/// Report offence to the `ReportOffence` handler.
+	fn report_evidence(
+		_reporter: Option<Reporter>,
+		_offence_proof: OffenceProof,
+		_key_owner_proof: Self::KeyOwnerProof,
+	) -> DispatchResult {
+		Ok(())
+	}
+
+	/// Check if is a known offence.
+	fn check_evidence(
+		_offence_proof: &OffenceProof,
+		_key_owner_proof: &Self::KeyOwnerProof,
+	) -> DispatchResult {
+		Ok(())
```

Maybe this should return an error for spam prevention?